File: 2021/day5/n.js
const text = await Bun.file("input").text();
const input = text
  .split("\n")
  .filter(Boolean)
  .map((line) => line.split(" -> ").map((part) => part.split(",").map(Number)));

const positions = {};

const addToPos = (x, y) => {
  const key = `${x},${y}`;
  if (!positions[key]) {
    positions[key] = 0;
  }
  positions[key] = positions[key] + 1;
};

input.forEach(([[x1, y1], [x2, y2]]) => {
  if (x1 === x2) {
    const min = Math.min(y1, y2);
    const max = Math.max(y1, y2);
    for (let i = min; i <= max; i++) {
      addToPos(x1, i);
    }
  }
  if (y1 === y2) {
    const min = Math.min(x1, x2);
    const max = Math.max(x1, x2);
    for (let i = min; i <= max; i++) {
      addToPos(i, y1);
    }
  }
});

const res = Object.values(positions).filter((v) => v > 1).length;
console.log(res);

File: 2021/day7/n.go
package main

import (
	"math"
	"strconv"
	"strings"
)

func parseInput(input string) []int {
	nums := strings.Split(strings.TrimSpace(input), ",")
	numbers := make([]int, 0, len(nums))

	for _, v := range nums {
		num, _ := strconv.Atoi(v)
		numbers = append(numbers, num)
	}

	return numbers
}

func min(nums []int) int {
	minVal := nums[0]
	for _, v := range nums {
		if v < minVal {
			minVal = v
		}
	}
	return minVal
}

func max(nums []int) int {
	maxVal := nums[0]
	for _, v := range nums {
		if v > maxVal {
			maxVal = v
		}
	}
	return maxVal
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

func SolveFirst(input string) int {
	positions := parseInput(input)
	minFuel := math.MaxInt
	minPos := min(positions)
	maxPos := max(positions)

	for i := minPos; i <= maxPos; i++ {
		cost := 0
		for _, pos := range positions {
			cost += abs(pos - i)
		}
		if cost < minFuel {
			minFuel = cost
		}
	}

	return minFuel
}

func SolveSecond(input string) int {
	positions := parseInput(input)
	minFuel := math.MaxInt
	minPos := min(positions)
	maxPos := max(positions)
	costs := make([]int, maxPos+1)

	for i := 1; i <= maxPos; i++ {
		costs[i] = costs[i-1] + i
	}

	for i := minPos; i <= maxPos; i++ {
		cost := 0
		for _, pos := range positions {
			delta := abs(pos - i)
			cost += costs[delta]
		}
		if cost < minFuel {
			minFuel = cost
		}
	}

	return minFuel
}
